src/core/server/http/base_path_proxy_server.ts

```diff
@@ -0,0 +1,169 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Server } from 'hapi-latest';
+import { Agent as HttpsAgent, ServerOptions as TlsOptions } from 'https';
+import { sample } from 'lodash';
+import { ByteSizeValue } from '../config/schema';
+import { DevConfig } from '../dev';
+import { Logger } from '../logging';
+import { HttpConfig } from './http_config';
+import { createServer, getServerOptions } from './http_tools';
+
+const alphabet = 'abcdefghijklmnopqrztuvwxyz'.split('');
+
+export interface BasePathProxyServerOptions {
+  httpConfig: HttpConfig;
+  devConfig: DevConfig;
+  isKibanaPath: (path: string) => boolean;
+  blockUntil: () => Promise<void>;
+}
+
+export class BasePathProxyServer {
+  private server?: Server;
+  private httpsAgent?: HttpsAgent;
+
+  get basePath() {
+    return this.options.httpConfig.basePath;
+  }
+
+  get targetPort() {
+    return this.options.devConfig.basePathProxyTargetPort;
+  }
+
+  constructor(
+    private readonly log: Logger,
+    private readonly options: BasePathProxyServerOptions
+  ) {
+    const ONE_GIGABYTE = 1024 * 1024 * 1024;
+    options.httpConfig.maxPayload = new ByteSizeValue(ONE_GIGABYTE);
+
+    if (!options.httpConfig.basePath) {
+      options.httpConfig.basePath = `/${sample(alphabet, 3).join('')}`;
+    }
+  }
+
+  public async start() {
+    const { httpConfig } = this.options;
+
+    const options = getServerOptions(httpConfig);
+    this.server = createServer(options);
+
+    // Register hapi plugin that adds proxying functionality. It can be configured
+    // through the route configuration object (see { handler: { proxy: ... } }).
+    await this.server.register({ plugin: require('h2o2-latest') });
+
+    if (httpConfig.ssl.enabled) {
+      const tlsOptions = options.tls as TlsOptions;
+      this.httpsAgent = new HttpsAgent({
+        ca: tlsOptions.ca,
+        cert: tlsOptions.cert,
+        key: tlsOptions.key,
+        passphrase: tlsOptions.passphrase,
+        rejectUnauthorized: false,
+      });
+    }
+
+    this.setupRoutes();
+
+    this.log.info(
+      `starting basepath proxy server at ${this.server.info.uri}${
+        httpConfig.basePath
+      }`
+    );
+
+    await this.server.start();
+  }
+
+  public async stop() {
+    this.log.info('stopping basepath proxy server');
+
+    if (this.server !== undefined) {
+      await this.server.stop();
+      this.server = undefined;
+    }
+
+    if (this.httpsAgent !== undefined) {
+      this.httpsAgent.destroy();
+      this.httpsAgent = undefined;
+    }
+  }
+
+  private setupRoutes() {
+    if (this.server === undefined) {
```

Comment:
I agree that these `=== undefined` checks aren't nice, but defining routes inside `start` cluttered this metod a bit, so during review we agreed to moved this to a dedicated method. "Static" helper will need additional `httpsAgent` parameter (or create it inside and return to `BasePathProxyServer` so that we can destroy it once server is stopped).

None of the alternatives stands out to be honest, and that's really an ergonomics issue with the approach we use across many services/servers inside new platform: we lazily create stuff (e.g. `server`) inside of `start` method, right before we want to use it so that we can asynchronously (via `Observable`) grab the latest configuration object. And we can't put everything into `start` in every case.

So I'm going to keep this as is for now if you don't mind but obviously we should discuss this "pattern" at some point and find a trade off so that we can have a ready recommendation for other Kibana folks once they start interacting with new platform.